apps/server-asset-sg/src/features/files/files.controller.ts

```diff
@@ -57,41 +65,73 @@ export class FilesController {
 
   @Post('/')
   @UseInterceptors(FileInterceptor('file', { limits: { fileSize: 250 * 1024 * 1024 } }))
-  async upload(@Req() req: Request, @UploadedFile() file: Express.Multer.File, @CurrentUser() user: User) {
-    const assetId = parseInt((req.body as { assetId?: string }).assetId ?? '');
-    if (isNaN(assetId)) {
-      throw new HttpException('missing assetId', HttpStatus.BAD_REQUEST);
-    }
+  async upload(
+    @Param('assetId', ParseIntPipe) assetId: number,
+    @Req() req: Request,
+    @UploadedFile() file: Express.Multer.File,
+    @CurrentUser() user: User
+  ) {
     const asset = await this.assetEditRepo.find(assetId);
     if (asset == null) {
       throw new HttpException('not found', HttpStatus.NOT_FOUND);
     }
     authorize(AssetEditPolicy, user).canUpdate(asset);
 
-    const result = await this.assetEditService.uploadFile(user, asset.assetId, {
+    const type = pipe(
+      AssetFileType.decode((req.body as { type?: string }).type ?? ''),
+      E.getOrElseW(() => null)
+    );
+    if (type == null) {
+      throw new HttpException('invalid type', HttpStatus.BAD_REQUEST);
+    }
+
+    const legalDocItemCode = pipe(
+      D.nullable(LegalDocItemCode).decode((req.body as { legalDocItemCode?: string }).legalDocItemCode ?? null),
+      E.getOrElseW(() => false as const)
+    );
+    if (legalDocItemCode === false) {
+      throw new HttpException('invalid legalDocItemCode', HttpStatus.BAD_REQUEST);
+    }
+    switch (type) {
```

Comment:
These kind of rules look like standard validation for me - we have the package 'class-validator' which is also recommended by nestjs, is this applicable here?

Reply:
Yes and no - `class-validator` could validate this. However, the asset-files are part of the `AssetDetail`/`AssetEdit` types, which are the only part of Assets that is still using `fp-ts`/`io-ts` for validation. So to check these correctly, we would have to move all asset things to `class-validator`, which we have so far avoided.

For now, I can rewrite the `FilesController` to use `class-validator`, but everywhere else, we're stuck with the current types.